### ISSUE TITLE
Custom replacement policy

### ISSUE THREAD

**Issue by freva** [vdurmont/emoji-java#27]
Currently there is no (simple) way to customize the replacement of unicode emojis, for example, if someone wants to change the prefix and suffixes used by `parseToAlias()`, one would actually have to not only re-implement the entire `parseToAlias()`, but also copy over the `EmojiTrie `(because the `Matches `enum inside it is package-private). 

I propose to define `EmojiTransformer `interface that requires a `String transform(UnicodeCandidate unicodeCandidate)`. Then make a general transformation method:

```java
    public static String parseFromUnicode(String input, EmojiTransformer transformer) {
        int prev = 0;
        StringBuilder sb = new StringBuilder();
        List<UnicodeCandidate> replacements = getUnicodeCandidates(input);
        for(UnicodeCandidate candidate : replacements) {
            sb.append(input.substring(prev, candidate.startIndex));

            sb.append(transformer.transform(candidate));
            prev = candidate.getEndIndexWithFitzpatrick();
        }

        return sb.append(input.substring(prev)).toString();
    }
```

Now any transformation can performed by sending in `EmojiTransformer` instance. As a proof-of-concept, I changed all from-unicode parsers in EmojiParser to use this model:

```java
    public static String parseToAliases(String input, final FitzpatrickAction fitzpatrickAction) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                switch (fitzpatrickAction) {
                    default:
                    case PARSE:
                        if(unicodeCandidate.hasFitzpatrick()) {
                            return ":" + unicodeCandidate.getEmoji().getAliases().get(0) + "|" +
                                    unicodeCandidate.getFitzpatrickType() + ":";
                        }

                    case REMOVE:
                    return ":" + unicodeCandidate.getEmoji().getAliases().get(0) + ":";

                    case IGNORE:
                    return ":" + unicodeCandidate.getEmoji().getAliases().get(0) + ":" +
                                unicodeCandidate.getFitzpatrickUnicode();

                }
            }
        };

        return parseFromUnicode(input, emojiTransformer);
    }


    public static String parseToHtmlDecimal(String input, final FitzpatrickAction fitzpatrickAction) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                switch (fitzpatrickAction) {
                    default:
                    case PARSE:
                    case REMOVE:
                        return unicodeCandidate.getEmoji().getHtmlDecimal();

                    case IGNORE:
                        return unicodeCandidate.getEmoji().getHtmlDecimal() + unicodeCandidate.getFitzpatrickUnicode();
                }
            }
        };

        return parseFromUnicode(input, emojiTransformer);
    }


    public static String parseToHtmlHexadecimal(String input, final FitzpatrickAction fitzpatrickAction) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                switch (fitzpatrickAction) {
                    default:
                    case PARSE:
                    case REMOVE:
                        return unicodeCandidate.getEmoji().getHtmlHexidecimal();

                    case IGNORE:
                        return unicodeCandidate.getEmoji().getHtmlHexidecimal() + unicodeCandidate.getFitzpatrickUnicode();
                }
            }
        };

        return parseFromUnicode(input, emojiTransformer);
    }


    public static String removeAllEmojis(String str) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                return "";
            }
        };

        return parseFromUnicode(str, emojiTransformer);
    }


    public static String removeEmojis(String str, final Collection<Emoji> emojisToRemove) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                if (! emojisToRemove.contains(unicodeCandidate.getEmoji())) {
                    return unicodeCandidate.getEmoji().getUnicode() + unicodeCandidate.getFitzpatrickUnicode();
                }
                return "";
            }
        };

        return parseFromUnicode(str, emojiTransformer);
    }


    public static String removeAllEmojisExcept(String str, final Collection<Emoji> emojisToKeep) {
        EmojiTransformer emojiTransformer = new EmojiTransformer() {
            public String transform(UnicodeCandidate unicodeCandidate) {
                if (emojisToKeep.contains(unicodeCandidate.getEmoji())) {
                    return unicodeCandidate.getEmoji().getUnicode() + unicodeCandidate.getFitzpatrickUnicode();
                }
                return "";
            }
        };

        return parseFromUnicode(str, emojiTransformer);
    }
```

I then ran a quick performance test on this for fun, and for 10k tweets the results were inconclusive, for 1m tweets the new method was twice as fast, so I went all-in and tested with 82.5m tweets and got:

```
Took 433510ms with the new method against 1099736ms with the old method
```

Only disadvantage with this method is that it doesn't look super pretty, but in Java 8 i looks way nicer, for example:

```java
String parsed = EmojiParser.parseFromUnicode("test", e -> " ||" + e.getEmoji().getAliases().get(0) + "|| ");
```
Which replaces `Here is a boy: \uD83D\uDC66!` to `Here is a boy:  ||boy|| !`

I'm not sure if there are better ways of doing this, if you like this one, I can submit a pull request.

**Comment by vdurmont** [comment]
Hey,
Sorry for being slow! I'll take a look asap, it seems very promising!

**Comment by vdurmont** [comment]
It's awesome, thanks!
I just merged #29.
